### ISSUE TITLE
macro operation fusion: addi + sw / lw + addi

### ISSUE THREAD

**Issue by jserv** [sysprog21/rv32emu#218]
The current macro operation fusion is capable of transforming multiple `LW` and `SW` instructions into a single fused instruction. However, it has some clever optimizations.

Consider the following instruction sequence:
```
   108e4:       fd010113                addi    sp,sp,-48
   108e8:       01312e23                sw      s3,28(sp)
   108ec:       02112623                sw      ra,44(sp)
   108f0:       02812423                sw      s0,40(sp)
   108f4:       02912223                sw      s1,36(sp)
   108f8:       03212023                sw      s2,32(sp)
   108fc:       01412c23                sw      s4,24(sp)
   10900:       01512a23                sw      s5,20(sp)
   10904:       01612823                sw      s6,16(sp)
   10908:       01712623                sw      s7,12(sp)
   1090c:       01812423                sw      s8,8(sp)
   10910:       01912223                sw      s9,4(sp)
   10914:       01a12023                sw      s10,0(sp)
```

and
```
   10b24:       02c12083                lw      ra,44(sp)
   10b28:       02812403                lw      s0,40(sp)
   10b2c:       02412483                lw      s1,36(sp)
   10b30:       02012903                lw      s2,32(sp)
   10b34:       01c12983                lw      s3,28(sp)
   10b38:       01812a03                lw      s4,24(sp)
   10b3c:       01412a83                lw      s5,20(sp)
   10b40:       01012b03                lw      s6,16(sp)
   10b44:       00c12b83                lw      s7,12(sp)
   10b48:       00812c03                lw      s8,8(sp)
   10b4c:       00412c83                lw      s9,4(sp)
   10b50:       00012d03                lw      s10,0(sp)
   10b54:       03010113                addi    sp,sp,48
```

The two instruction sequences always appear in a specific order. At the beginning of the function, the `SW` sequence is determined by the immediate value of the `ADDI` instruction. Similarly, after multiple `LW` instructions, an `ADDI` instruction follows with an explicit immediate value within a certain range. We can fuse these sequences without passing or manipulating addresses since we can always calculate them.

In other words, we introduce a fast path for the function prologue and epilogue for translating C (or other high-level programming languages) into RISC-V instructions with the proper calling convention.

**Comment by qwe661234** [comment]
With PR #240, we can fuse all sequential lw and sw even if they has been reordered by compiler. I think leave an addi unfused is acceptable.
